File: LCC/Programacion2/tp/test_cases/test_othello.c

#include <assert.h>
#include <stdio.h>
#include <string.h>

// --- estructuras (del othello.c) ---

typedef struct {
    char *nombre;
    char color;
} Jugador;

typedef struct {
    Jugador j1;
    Jugador j2;
    char turno;
} Juego;

typedef struct {
    int r;
    int c;
} Pos;

int dentro(int r, int c);
char otro_color(char color);
Pos obtener_direccion(int dir);
void inicializar_tablero(char board[8][8]);

int parse_move(const char *move, int *out_r, int *out_c);
int calcular_flips(char board[8][8], char my, int r, int c, Pos flips[64]);
int hay_alguna_jugada_valida(char board[8][8], char color);

void aplicar_jugada(char board[8][8], char my, int r, int c, Pos flips[64], int cant_flips);
int verificar_ganador(char board[8][8]);

// --- helpers ---

void tablero_vacio(char board[8][8]) {
    for (int r = 0; r < 8; ++r)
        for (int c = 0; c < 8; ++c)
            board[r][c] = 'X';
}

void tablero_inicial_estandar(char board[8][8]) {
    // el init clasico del juego
    tablero_vacio(board);
    board[4][3] = 'N'; // D5
    board[4][4] = 'B'; // E5
    board[3][4] = 'N'; // E4
    board[3][3] = 'B'; // D4
}

int mismo_set_flips(Pos a[64], int na, Pos b[64], int nb) {
    // para comparar sin importar el orden
    if (na != nb) return 0;

    for (int i = 0; i < na; ++i) {
        int ok = 0;
        for (int j = 0; j < nb; ++j) {
            if (a[i].r == b[j].r && a[i].c == b[j].c) {
                ok = 1;
                break;
            }
        }
        if (!ok) return 0;
    }
    return 1;
}

// --- tests unitarios ---

void test_obtener_direccion() {
    Pos p;

    // dir 0: (-1, -1)
    p = obtener_direccion(0);
    assert(p.r == -1);
    assert(p.c == -1);

    // dir 3: (0, -1)
    p = obtener_direccion(3);
    assert(p.r == 0);
    assert(p.c == -1);

    // dir 7: (1, 1)
    p = obtener_direccion(7);
    assert(p.r == 1);
    assert(p.c == 1);

    // fuera de rango -> (0,0)
    p = obtener_direccion(-1);
    assert(p.r == 0);
    assert(p.c == 0);

    p = obtener_direccion(8);
    assert(p.r == 0);
    assert(p.c == 0);
}

void test_dentro_y_otro_color() {
    assert(dentro(0, 0) == 1);
    assert(dentro(7, 7) == 1);
    assert(dentro(-1, 0) == 0);
    assert(dentro(0, -1) == 0);
    assert(dentro(8, 0) == 0);
    assert(dentro(0, 8) == 0);

    assert(otro_color('B') == 'N');
    assert(otro_color('N') == 'B');
}

void test_parse_move() {
    int r, c;

    r = c = -1;
    assert(parse_move("D6", &r, &c) == 1);
    assert(r == 5 && c == 3);

    r = c = -1;
    assert(parse_move("A1", &r, &c) == 1);
    assert(r == 0 && c == 0);

    r = c = -1;
    assert(parse_move("", &r, &c) == 2); // pass
    assert(r == -1 && c == -1); // no deberia tocarlo

    r = c = -1;
    assert(parse_move("AA", &r, &c) == 0);
    assert(parse_move("A9", &r, &c) == 0);
    assert(parse_move("I1", &r, &c) == 0);
    assert(parse_move("D66", &r, &c) == 0);
}

void test_calcular_flips_en_inicio() {
    char board[8][8];
    tablero_inicial_estandar(board);

    // en el tablero inicial, B tiene 4 jugadas validas y en todas voltea 1
    // en particular, C5 (r=4,c=2) voltea D5 (r=4,c=3)
    Pos flips[64];
    int cant = calcular_flips(board, 'B', 4, 2, flips);
    assert(cant == 1);
    assert(flips[0].r == 4 && flips[0].c == 3);

    // y una que no deberia ser valida
    cant = calcular_flips(board, 'B', 0, 0, flips);
    assert(cant == 0);

    // celda ocupada -> 0
    cant = calcular_flips(board, 'B', 3, 3, flips);
    assert(cant == 0);
}

void test_hay_alguna_jugada_valida() {
    char board[8][8];
    tablero_inicial_estandar(board);

    assert(hay_alguna_jugada_valida(board, 'B') == 1);
    assert(hay_alguna_jugada_valida(board, 'N') == 1);

    // tablero casi todo B, con una X y sin N: nadie puede encerrar nada
    tablero_vacio(board);
    for (int r = 0; r < 8; ++r)
        for (int c = 0; c < 8; ++c)
            board[r][c] = 'B';
    board[0][0] = 'X';

    assert(hay_alguna_jugada_valida(board, 'B') == 0);
    assert(hay_alguna_jugada_valida(board, 'N') == 0);
}

void test_aplicar_jugada_voltea() {
    char board[8][8];
    tablero_inicial_estandar(board);

    // probamos una jugada concreta: B juega C5 (4,2) y voltea D5 (4,3)
    Pos flips[64];
    int cant = calcular_flips(board, 'B', 4, 2, flips);
    assert(cant == 1);

    aplicar_jugada(board, 'B', 4, 2, flips, cant);

    assert(board[4][2] == 'B');
    assert(board[4][3] == 'B'); // volteada
}

void test_calcular_flips_mas_de_uno() {
    // armo un caso facil donde una jugada voltea 3 en linea
    // fila 4: X N N N B ...
    // B juega A4 -> voltea B4,C4,D4
    char board[8][8];
    tablero_vacio(board);

    board[3][1] = 'N';
    board[3][2] = 'N';
    board[3][3] = 'N';
    board[3][4] = 'B';

    Pos flips[64];
    int cant = calcular_flips(board, 'B', 3, 0, flips);
    assert(cant == 3);

    Pos esperado[64];
    esperado[0].r = 3; esperado[0].c = 1;
    esperado[1].r = 3; esperado[1].c = 2;
    esperado[2].r = 3; esperado[2].c = 3;

    assert(mismo_set_flips(flips, cant, esperado, 3) == 1);
}

void test_verificar_ganador() {
    char board[8][8];

    // no terminado si hay alguna X
    tablero_vacio(board);
    board[0][0] = 'B';
    assert(verificar_ganador(board) == 0);

    // terminado si no hay X
    for (int r = 0; r < 8; ++r)
        for (int c = 0; c < 8; ++c)
            board[r][c] = 'B';
    assert(verificar_ganador(board) == 1);
}

// --- runner ---

int main() {
    test_dentro_y_otro_color();
    test_obtener_direccion();
    test_parse_move();
    test_calcular_flips_en_inicio();
    test_hay_alguna_jugada_valida();
    test_aplicar_jugada_voltea();
    test_calcular_flips_mas_de_uno();
    test_verificar_ganador();

    printf("ok, pasaron todos los tests\n");
    return 0;
}

File: LCC/Programacion2/tp/test_cases/test_othello.py
import pytest
import othello

def _tablero_vacio():
    return othello.inicializar_tablero()

def _poner(tablero, coords, ch):
    # helper
    for r, c in coords:
        tablero[r][c] = ch
    return tablero

def _tablero_inicial_estandar():
    # setup clasico del othello en el centro
    t = _tablero_vacio()
    _poner(t, [(3, 3), (4, 4)], 'B')
    _poner(t, [(3, 4), (4, 3)], 'N')
    return t

def test_inicializar_tablero():
    t = othello.inicializar_tablero()
    assert len(t) == 8
    assert all(len(fila) == 8 for fila in t)
    assert all(t[r][c] == 'X' for r in range(8) for c in range(8))

def test_otro_color():
    assert othello.color_opuesto('B') == 'N'
    assert othello.color_opuesto('N') == 'B'

def test_parsear_jugada_ok():
    assert othello.parsear_jugada("A1") == (0, 0)
    assert othello.parsear_jugada("H8") == (7, 7)
    assert othello.parsear_jugada("D6") == (5, 3)
    assert othello.parsear_jugada("Y9") == None
    assert othello.parsear_jugada("D78") == None

def test_parsear_jugada_invalidas():
    # parsear_jugada asume mayus, asi que no le paso minusculas aca
    invalidas = ["", "A", "AA", "A9", "I1", "11", "Z0", "A0", "H9", "D66"]
    for txt in invalidas:
        assert othello.parsear_jugada(txt) is None

def test_dentro():
    assert othello.dentro(0, 0) is True
    assert othello.dentro(7, 7) is True
    assert othello.dentro(-1, 0) is False
    assert othello.dentro(0, -1) is False
    assert othello.dentro(8, 0) is False
    assert othello.dentro(0, 8) is False

def test_flips_de_jugada_en_tablero_inicial():
    t = _tablero_inicial_estandar()
    othello.imprimir_tablero(t)

    # en el tablero inicial, para B las jugadas validas son 4 y cada una voltea 1
    # ejemplo: C4 (r=3,c=2) voltea (3,3)
    flips = othello.flips_de_jugada(t, 'N', 3, 2)
    assert flips == [(3, 3)]

    flips = othello.flips_de_jugada(t, 'N', 2, 3)  # D3 voltea (3,3)
    assert flips == [(3, 3)]

    flips = othello.flips_de_jugada(t, 'B', 4, 5)  # F5 voltea (4,4)? no, al reves: voltea (4,4) si hay N.. aca no
    # mejor chequeo otra jugada valida de B: E6 (5,4) voltea (4,4)? tampoco
    # en realidad para B en inicio: C4 D3 E6 F5 (todas voltean 1)
    flips = othello.flips_de_jugada(t, 'B', 5, 4)  # E6 voltea (4,4) (que es B) nope, me equivoque si lo miro mal
    # para evitar confusiones, chequeamos las 4 validas via jugadas_validas mas abajo

def test_jugadas_validas_inicial_estandar():
    t = _tablero_inicial_estandar()

    movs_b = othello.jugadas_validas(t, 'B')
    movs_n = othello.jugadas_validas(t, 'N')

    # en inicio hay 4 jugadas validas por lado
    assert len(movs_b) == 4
    assert len(movs_n) == 4

    # y en todas se voltea 1
    assert all(len(flips) == 1 for flips in movs_b.values())
    assert all(len(flips) == 1 for flips in movs_n.values())

def test_aplicar_jugada_voltea_correcto():
    t = _tablero_inicial_estandar()
    movs_b = othello.jugadas_validas(t, 'B')

    # elegimos una valida cualquiera, y aplicamos usando los flips ya calculados
    (r, c), flips = next(iter(movs_b.items()))
    othello.aplicar_jugada(t, 'B', r, c, flips)

    assert t[r][c] == 'B'
    for x, y in flips:
        assert t[x][y] == 'B'

def test_elegir_jugada_maquina_level0_y_level1():
    t = _tablero_inicial_estandar()
    movs = othello.jugadas_validas(t, 'B')

    # lvl 0 devuelve alguna valida
    pick0 = othello.elegir_jugada_maquina(movs, 0)
    assert pick0 is not None
    assert pick0[0] in movs

    # lvl 1 en inicio da igual cual, todas voltean 1. pero tiene que ser valida.
    pick1 = othello.elegir_jugada_maquina(movs, 1)
    assert pick1 is not None
    assert pick1[0] in movs
    assert len(pick1[1]) == 1

def test_tablero_lleno():
    t = _tablero_vacio()
    assert othello.tablero_lleno(t) is False

    _poner(t, [(r, c) for r in range(8) for c in range(8)], 'B')
    assert othello.tablero_lleno(t) is True

def test_contar_fichas():
    t = _tablero_vacio()
    _poner(t, [(0, 0), (0, 1), (0, 2)], 'B')
    _poner(t, [(7, 7), (6, 6)], 'N')
    bc, nc = othello.contar_fichas(t)
    assert bc == 3
    assert nc == 2

def test_juego_terminado_por_tablero_lleno():
    t = _tablero_vacio()
    _poner(t, [(r, c) for r in range(8) for c in range(8)], 'B')

    # turno da igual, tablero lleno termina
    assert othello.juego_terminado(t, 'B') is True
    assert othello.juego_terminado(t, 'N') is True

def test_juego_terminado_sin_jugadas_para_ninguno():
    # armamos un tablero lleno de B con UNA X suelta, pero sin N en ningun lado
    # entonces no hay jugadas validas para nadie (pq no se puede encerrar nada)
    t = _tablero_vacio()
    _poner(t, [(r, c) for r in range(8) for c in range(8)], 'B')
    t[0][0] = 'X'

    assert othello.jugadas_validas(t, 'B') == {}
    assert othello.jugadas_validas(t, 'N') == {}
    assert othello.juego_terminado(t, 'B') is True

def test_cargar_partida_desde_archivo_ok(tmp_path):
    # uso tmp_path asi no ensuciamos el repo
    contenido = [
        "XXXXXXXX",
        "XXXXXXXX",
        "XXXXXXXX",
        "XXXBNXXX",
        "XXXNBXXX",
        "XXXXXXXX",
        "XXXXXXXX",
        "XXXXXXXX",
        "B",
    ]
    p = tmp_path / "partida.txt"
    p.write_text("\n".join(contenido), encoding="utf-8")

    tablero, turno = othello.cargar_partida_desde_archivo(str(p))
    assert turno == 'B'
    assert tablero[3][3] == 'B'
    assert tablero[3][4] == 'N'
    assert tablero[4][3] == 'N'
    assert tablero[4][4] == 'B'

def test_cargar_partida_desde_archivo_invalido_menos_lineas(tmp_path):
    p = tmp_path / "bad.txt"
    p.write_text("\n".join(["XXXXXXXX"] * 8), encoding="utf-8")
    with pytest.raises(Exception):
        othello.cargar_partida_desde_archivo(str(p))

def test_cargar_partida_desde_archivo_invalido_fila_corta(tmp_path):
    contenido = ["XXXXXXX"] + ["XXXXXXXX"] * 7 + ["B"]
    p = tmp_path / "bad.txt"
    p.write_text("\n".join(contenido), encoding="utf-8")
    with pytest.raises(Exception):
        othello.cargar_partida_desde_archivo(str(p))

def test_cargar_partida_desde_archivo_invalido_char_raro(tmp_path):
    contenido = ["XXXXXXXX"] * 3 + ["XXXQNXXX"] + ["XXXXXXXX"] * 4 + ["B"]
    p = tmp_path / "bad.txt"
    p.write_text("\n".join(contenido), encoding="utf-8")
    with pytest.raises(Exception):
        othello.cargar_partida_desde_archivo(str(p))

def test_cargar_partida_desde_archivo_invalido_turno(tmp_path):
    contenido = ["XXXXXXXX"] * 8 + ["Z"]
    p = tmp_path / "bad.txt"
    p.write_text("\n".join(contenido), encoding="utf-8")
    with pytest.raises(Exception):
        othello.cargar_partida_desde_archivo(str(p))
